src/common/os_input_output.rs

```diff
@@ -156,24 +165,53 @@ fn spawn_terminal(file_to_open: Option<PathBuf>, orig_termios: termios::Termios)
     (pid_primary, pid_secondary)
 }
 
+/// Sends messages on an [ipmpsc](ipmpsc) channel, along with an [`ErrorContext`].
 #[derive(Clone)]
-pub struct OsInputOutput {
+struct IpcSenderWithContext<T: Serialize> {
+    err_ctx: ErrorContext,
+    sender: IpcSender,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Serialize> IpcSenderWithContext<T> {
+    /// Returns a sender to the given [SharedRingBuffer](ipmpsc::SharedRingBuffer).
+    fn new(buffer: SharedRingBuffer) -> Self {
+        Self {
+            err_ctx: ErrorContext::new(),
+            sender: IpcSender::new(buffer),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Updates this [`IpcSenderWithContext`]'s [`ErrorContext`]. This is the way one adds
+    /// a call to the error context.
+    ///
+    /// Updating [`ErrorContext`]s works in this way so that these contexts are only ever
+    /// allocated on the stack (which is thread-specific), and not on the heap.
+    fn update(&mut self, ctx: ErrorContext) {
+        self.err_ctx = ctx;
+    }
+
+    /// Sends an event, along with the current [`ErrorContext`], on this
+    /// [`IpcSenderWithContext`]'s channel.
+    fn send(&mut self, msg: T) -> ipmpsc::Result<()> {
+        self.sender.send(&(msg, self.err_ctx))
+    }
+}
+
+#[derive(Clone)]
+pub struct ServerOsInputOutput {
     orig_termios: Arc<Mutex<termios::Termios>>,
+    server_sender: IpcSenderWithContext<ServerInstruction>,
+    server_receiver: Arc<Mutex<IpcReceiver>>,
```

Comment:
I was in quite a dilemma on how to store this. We require the clones of `os_input` across threads. I had two choices-
1. Have an `Arc<Mutex<IpcReceiver>>`. This enables `os_input` to be cloned and since the receiver is ideally used by a single thread (`server_router`), the lock resolves immediately every time.
2. Have an `Option<Receiver>`. Here the original `os_input` returned by `get_server_os_input()` would contain `Some(receiver)` and a custom `Clone` implementation would make all the clones created store `None` there. The original `os_input` would then be owned by the `server_router` thread.

I don't have a strong opinion about either. If 2 sounds like a better solution, I can change it.